### ISSUE TITLE
RFC - 6.0 - Don't nest input inside label by default for checkboxes/radios

### ISSUE THREAD

**Issue by ADmad** [cakephp/cakephp#18083]
### Description

IIRC we started the nesting by default as that's what Bootstrap did (the dominant CSS lib at the time).

As of Bootstrap 5.0 it no longer expects inputs to be nested inside the label tag. The UI world's current darling, Tailwind CSS, also doesn't seem to expect checkboxes/radios nested inside labels.

So this default nesting adds additional burden on the devs to remove, if they are using one of the current popular CSS libs. Hence my proposal to remove the default nesting and keep it as an optional feature.

### CakePHP Version

6.0

**Comment by LordSimal** [comment]
> The UI world's current darling, Tailwind CSS, also doesn't seem to expect checkboxes/radios nested inside labels.

Depends on what tailwind based package you use. [Flowbite](https://flowbite.com/docs/forms/checkbox/) doesn't nest but [DaisyUI](https://daisyui.com/components/checkbox/) does 

---

The major problem related to proper tailwind integration is the fact, that we have CSS classes as PHP props or config which tailwind can't detect easily.

I believe one could write custom regex patterns to make that work with our current way of templating `Helpers <=> Widgets <=> String Template`

It would be easier for tailwind to detect all necessary classes if they are all present inside a `class=""` attribute [for optimized production builds](https://tailwindcss.com/docs/optimizing-for-production#writing-purgeable-html). See e.g. [this laravel example](https://stackoverflow.com/questions/68909968/laravel-conditional-tailwind-class-with-default-value)

But thats a major view rewrite as far as I can tell.

**Comment by ADmad** [comment]
> Depends on what tailwind based package you use. [Flowbite](https://flowbite.com/docs/forms/checkbox/) doesn't nest but [DaisyUI](https://daisyui.com/components/checkbox/) does

I only looked at the examples here  https://tailwindui.com/components/application-ui/forms/form-layouts. Had no idea there are various UI libs built on top.

> The major problem related to proper tailwind integration is the fact, that we have CSS classes as PHP props or config which tailwind can't detect easily.

That's a separate topic of discussion.

**Comment by LordSimal** [comment]
FYI: Tailwind UI is one of many Tailwind based component libraries 😉

Tailwind itself is just the tool, that maps 1 HTML class to 1 CSS prop/feature/behavior

**Comment by markstory** [comment]
I have some memory of there being accessibility reasons for nesting as well. However, looking at the current status of how to build an accessible checkbox the only advantage I can find for nesting is that you don't need id attributes. However, we already generate those.

We could introduce this as an opt-in feature for 5.x and then flip the default in 6.x. That would give people a way to do their upgrade incrementally.

**Comment by ADmad** [comment]
> We could introduce this as an opt-in feature for 5.x and then flip the default in 6.x. 

Disabling the nesting for individual checkboxes is already possible in 5.x, requires setting the `nestedInput` option to `false` and customizing the `checkboxFormGroup` template. So it's a matter of switching the defaults in 6.x.

For multi checkboxes and radio it requires overriding their widget config to use `label` instead of `nestingLabel` and tweaking their form group templates.

**Comment by markstory** [comment]
Should we have a config option to change the default value of `nestedInput`? It would be simpler to update styles and flip one config option over having to change each checkbox control.

**Comment by ADmad** [comment]
> Should we have a config option to change the default value of nestedInput?

Yes I was thinking along those line too. Also the `nestedInput` option only affects single checkboxes. To disable the nesting for multi checkoxes and radio requires changing their respective widget config, passing in the `label` widget instead of `nestingLabel`.

**Comment by dereuromark** [comment]
Looks like with PR merged this is done.

**Comment by ADmad** [comment]
Changing the default to un-nested in 6.0 is pending.
